spring-cloud-stream/src/test/java/org/springframework/cloud/stream/function/SourceToFunctionsSupportTests.java

```diff
@@ -0,0 +1,226 @@
+		assertThat(target.receive(10000).getPayload())
+				.isEqualTo("11".getBytes(StandardCharsets.UTF_8));
+		assertThat(target.receive(10000).getPayload())
+				.isEqualTo("22".getBytes(StandardCharsets.UTF_8));
+		assertThat(target.receive(10000).getPayload())
+				.isEqualTo("33".getBytes(StandardCharsets.UTF_8));
+		//etc
+	}
+
+	@Test
+	public void testMessageSourceIsCreatedFromProvidedSupplierComposedWithMultipleFunctions() {
+		ApplicationContext context =
+				new SpringApplicationBuilder(
+						TestChannelBinderConfiguration.getCompleteConfiguration(SupplierConfiguration.class))
+						.web(WebApplicationType.NONE)
+						.run("--spring.cloud.stream.function.name=number|concatWithSelf|mmultiplyByTwo",
+								"--spring.jmx.enabled=false");
+
+		OutputDestination target = context.getBean(OutputDestination.class);
+		assertThat(target.receive(10000).getPayload())
+				.isEqualTo("22".getBytes(StandardCharsets.UTF_8));
+		assertThat(target.receive(10000).getPayload())
+				.isEqualTo("44".getBytes(StandardCharsets.UTF_8));
+		assertThat(target.receive(10000).getPayload())
+				.isEqualTo("66".getBytes(StandardCharsets.UTF_8));
+		//etc
+	}
+
+	@EnableAutoConfiguration
+	@Import(ProvidedMessageSourceConfiguration.class)
+	public static class SupplierConfiguration {
+		AtomicInteger counter = new AtomicInteger();
+		@Bean
+		public Supplier<String> number() {
+			return () -> String.valueOf(counter.incrementAndGet());
+		}
+
+		@Bean
+		public Function<String, String> concatWithSelf() {
+			return x -> x + x;
+		}
+
+		@Bean
+		public Function<Flux<String>, Flux<String>> mmultiplyByTwo() {
+			return x -> x.map(i -> String.valueOf(Integer.valueOf(i) * 2));
+		}
+	}
+
+	@EnableAutoConfiguration
+	@Import(ExistingMessageSourceConfiguration.class)
+	public static class FunctionsConfiguration {
+
+		@Bean
+		public Function<String, String> toUpperCase() {
+			return String::toUpperCase;
+		}
+
+		@Bean
+		public Function<String, String> concatWithSelf() {
+			return x -> x + ":" + x;
+		}
+	}
+
+	/**
+	 * This configuration essentially emulates our existing app-starters for Sources
+	 * and essentially demonstrates how a function(s) could be applied to an existing
+	 * source via {@link IntegrationFlowFunctionSupport} class.
+	 */
+	@EnableBinding(Source.class)
+	public static class ExistingMessageSourceConfiguration {
+
+		@Autowired
+		private Source source;
+
+		@Bean
+		public IntegrationFlow messageSourceFlow(IntegrationFlowFunctionSupport functionSupport) {
+			Supplier<Message<String>> messageSource = () -> MessageBuilder.withPayload("hello function")
+					.setHeader(MessageHeaders.CONTENT_TYPE, MimeTypeUtils.TEXT_PLAIN).build();
+
+			IntegrationFlowBuilder flowBuilder = functionSupport.integrationFlowFromProvidedSupplier(messageSource);
+
+			if (!functionSupport.andThenFunction(flowBuilder, source.output())) {
+				flowBuilder = flowBuilder.channel(source.output());
```

Comment:
I don't like this combination of operators, but this is not your fault. This is a leak of proper reactive support in SI...
Hope we will fix it eventually. For example seems for me I suggested in some our discussion something like `fluxTransform(Fucntion<Flux<?>, Flux<?>>)` for the current Java DSL. What you would need to do here is just a function resolution from auto-configuration. 

Maybe it's really a time to play with that there?..

Reply:
Yes, consider opening appropriate INT issues, so once addressed we can simplify the code here.